### PR TITLE
Add docker definition for Mbed OS.

### PR DESCRIPTION
**Pull request by lukgni**
#### Problem
Missing Mbed OS sources, tools and toolchain required to start working on the Mbed OS <-> CHIP integration.

 #### Summary of Changes
- Add Dockerfile for creating chip-build-mbedos
- Update chip-build-vscode/Dockerfile
- Update integration/images version file (0.4.18 -> 0.5.0)
- Add mbed-cli and mbed-tools to scripts/requirements.in
- Regenerate scripts/requirements.txt (pip-compile)
- Fix bug with unactive link to Device Tree Compiler in chip-build-nrf-platform Dockerfile.
- Move Cmake 3.19.3 installation from chip-build-mbedos to chip-build.
- Update docker images version to 0.5.0 (according to integrations/images/chip-build/README.md).
- Change used images version in devcontainer.json
- Limit 'wget' outputs during chip-build-mbedos image building.

 #### Testing
To test this change, please run **integration/docker/images/build-all.sh** script first. 
This will build all docker images locally. Then, please follow steps described in  [VSCODE_DEVELOPMENT.md](https://github.com/project-chip/connectedhomeip/blob/master/docs/VSCODE_DEVELOPMENT.md). Don't forget to run ./scripts/activate.sh!

If you will receive error during Post-vscode container build - don't worry, this bug has been already [solved](https://github.com/project-chip/connectedhomeip/pull/4404) in original repo.

### CODE REVIEW

File: integrations/docker/images/chip-build-mbedos/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    mercurial \
+    && rm -rf /var/lib/apt/lists \
+    && : # last line
+
+# ------------------------------------------------------------------------------
+# Install Mbed-OS sources
+RUN set -x \
+    && (mkdir -p /opt/mbed-os \
+    && cd /opt/mbed-os \
+    && git clone --progress -b mbed-os-6.6.0 https://github.com/ARMmbed/mbed-os.git ./) \
```
Comment (pan-) [inline integrations/docker/images/chip-build-mbedos/Dockerfile:23]: The git repository is quite large, would it be possible to download the release from github instead ?

File: integrations/docker/images/chip-build/version [hunk 2]
```diff
@@ -1 +1 @@
-0.4.18
+0.4.20
```
Comment (pan-) [inline integrations/docker/images/chip-build/version:1]: How does version number works ? Why the increase per 2.

**Review by pan- — APPROVED** [review]
Thanks @lukgni that looks great. I just wonder if we shouldn't download an archive of Mbed OS source code instead of cloning the repository to reduce the image size and speed its build.

File: integrations/docker/images/chip-build/version (hunk 2 above)
Comment (lukgni) [reply integrations/docker/images/chip-build/version:1]: Overall rules are defined [here](https://github.com/ARMmbed/connectedhomeip/tree/master/integrations/docker/images/chip-build).
Here is some [example](https://github.com/project-chip/connectedhomeip/commit/c6f2e84512e096e10b8f4660bd009795ace0ddc2#diff-087ab3369ff49004266cdc5efba580a58b98e1396c5062728d9717da5eb6532b).

Yep, you're right - we should increase version to 0.4.19 - my mistake.

File: integrations/docker/images/chip-build-mbedos/Dockerfile (hunk 1 above)
Comment (lukgni) [reply integrations/docker/images/chip-build-mbedos/Dockerfile:23]: Yes, I see now - we can download compressed sources (in .tar.gz) from github release page of Mbed OS.
Thanks, good point!

**Comment by lukgni** [comment]
I wonder how we should deal with multiple ARM toolchains inside chip-build-vscode image:

**integrations/docker/images/chip-build-vscode/Dockerfile**
```
COPY --from=nrf /opt/NordicSemiconductor/nRF5_tools /opt/NordicSemiconductor/nRF5_tools
COPY --from=nrf /opt/ARM-software/gcc-arm-none-eabi-9-2019-q4-major /opt/ARM-software/gcc-arm-none-eabi-9-2019-q4-major
COPY --from=nrf /opt/nrfconnect/sdk-nrf /opt/NordicSemiconductor/nrfconnect
...
COPY --from=mbedos /opt/mbed-os /opt/mbed-os
COPY --from=mbedos /opt/mbed-os-toolchain/ /opt/mbed-os-toolchain/
```
For now nrf and mbed-os containers uses the same version of gcc arm toolchain.
It would be good to reduces chip-build-vscode container size as much as possible and stay only with the one copy of the toolchain.

From other hand, that approach may cause some problems in the future, when for e.g. nrf or mbed-os would like to change used toolchain to the other version.

**Comment by pan-** [comment]
I think it is more important to have the correct dev tools installed than a small image. Having to separate toolchains in the developer image helps in that regard and avoid an implicit dependency between NRF and Mbed. 

That's not an image build very often and it isn't used by the CI so we should be fine.

File: integrations/docker/images/chip-build-mbedos/Dockerfile (hunk 1 above)
Comment (lukgni) [reply integrations/docker/images/chip-build-mbedos/Dockerfile]: Done.

File: integrations/docker/images/chip-build/version (hunk 2 above)
Comment (lukgni) [reply integrations/docker/images/chip-build/version]: Done.

File: integrations/docker/images/chip-build-mbedos/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    mercurial \
+    && rm -rf /var/lib/apt/lists \
+    && : # last line
+
+# ------------------------------------------------------------------------------
+# Install Mbed-OS sources
+RUN set -x \
+    && (mkdir -p /opt/mbed-os \
+    && cd /opt/mbed-os \
+    && wget https://github.com/ARMmbed/mbed-os/archive/mbed-os-6.6.0.tar.gz \
```
Comment (fkjagodzinski) [inline integrations/docker/images/chip-build-mbedos/Dockerfile:23]:
```suggestion
    && wget --progress=dot:giga https://github.com/ARMmbed/mbed-os/archive/mbed-os-6.6.0.tar.gz \
```
`wget` tends to produce tons of useless dotted progress output. You can limit that with the `giga` param.

**Review by fkjagodzinski — COMMENTED** [review]
I think we should also update the version in the `devcontainer.json` vscode file. The [README](https://github.com/project-chip/connectedhomeip/tree/master/integrations/docker/images/chip-build#docker-connectedhomeipchip-build) suggests that.

File: integrations/docker/images/chip-build/version [hunk 4]
```diff
@@ -1 +1 @@
-0.4.18
+0.4.19
```
[the remark this answers is not in the document]
Comment (lukgni) [reply integrations/docker/images/chip-build/version]: Done.

File: integrations/docker/images/chip-build-mbedos/Dockerfile (hunk 3 above)
Comment (lukgni) [reply integrations/docker/images/chip-build-mbedos/Dockerfile]: Done.

**Comment by lukgni** [comment]
> I think we should also update the version in the `devcontainer.json` vscode file. The [README](https://github.com/project-chip/connectedhomeip/tree/master/integrations/docker/images/chip-build#docker-connectedhomeipchip-build) suggests that.

Done.

File: integrations/docker/images/chip-build-nrf-platform/Dockerfile [hunk 5]
```diff
@@ -35,7 +35,7 @@ ENV ARM_GCC_INSTALL_ROOT=/opt/ARM-software/gcc-arm-none-eabi-9-2019-q4-major/bin
 # ==================================================
 RUN set -x \
     # Device Tree Compiler 1.4.7
-    && curl -o /tmp/dtc.deb http://mirrors.edge.kernel.org/ubuntu/pool/main/d/device-tree-compiler/device-tree-compiler_1.4.7-3ubuntu2_amd64.deb \
+    && curl -o /tmp/dtc.deb http://ftp.br.debian.org/debian/pool/main/d/device-tree-compiler/device-tree-compiler_1.4.7-3_amd64.deb \
```
Comment (fkjagodzinski) [inline integrations/docker/images/chip-build-nrf-platform/Dockerfile:38]: Nice to see this one fixed. :+1:

**Review by fkjagodzinski — APPROVED** [review]
LGTM. I couldn't test this as you described; I had to interrupt the build of `chip-build-esp32` -- submodule download took ages to complete. Eventually, I built and ran the Mbed container manually. Works OK. :+1: There are 2 things to keep in mind when we add vscode tasks for Mbed:
- add `/opt/mbed-os-toolchain/gcc-arm-none-eabi-9-2019-q4-major/bin` to `PATH` so that `mbedtools` CMake can use it,
- link `/opt/mbed-os` to project's dir for the `mbedtools` to use. I couldn't get this to work with `mbedtools compile --toolchain=GCC_ARM --mbed-target=K64F --mbed-os-path=/opt/mbed-os/`, same as you @lukgni.

Here are the steps I did for testing (may, or may not come in handy in the future):

```bash
# In connectedhomeip/integrations/docker/images/chip-build-mbedos
# Build with
./build.sh

# Run with
./run.sh -it -- bash

# In the container
cd $(mktemp -d)
git clone https://github.com/ARMmbed/mbed-os-example-blinky.git
cd mbed-os-example-blinky/
PATH="${MBED_GCC_ARM_PATH}:$PATH"
export PATH
which arm-none-eabi-gcc
ln -s ${MBED_OS_PATH} .
mbedtools compile --toolchain=GCC_ARM --mbed-target=K64F
ls cmake_build/mbed-os-example-blinky.hex
```

**Comment by ATmobica** [comment]
I agree with Filip that the mbed toolchain bin should be added to PATH. Mbed Cmake only defines compiler names, not full paths.  But adding a symbolic link to mbed-os source is not necessary. A few changes inside the project CMakeLists file which we already mentioned are enough to refer to the sources from /opt/.

**Comment by lukgni** [comment]
Ok, I updated dockerfile with PATH configuration.
With this change chip-build-mbedos should now work out-of-box.
 
To verify this change you can try to run container outside vscode, for e.g: (this will mount CHIP directory inside container)
`docker run  -it --mount type=bind,source=<CHIP absolute path>,target=/workspaces/connectedhomeip  <chip-build-mbedos ID>`

**Review by ATmobica — APPROVED** [review]

**Comment by lukgni** [comment]
Now I have to figure out how to deal with that pigweed tool which overrides cmake and other things. Until then we should wait with merging this PR (in current state) even if it will be approved by everyone.

**Review by pan- — APPROVED** [review]
Thanks for the PR @lukgni . 
Everything seems to work locally in the `chip-build-mbedos` image. The only hiccup I had - like everyone else -  was with using the option `--mbed-os-path` option which doesn't seems to work. 

As discussed on slack, I think it would be great to make a submission on the upstream repo so we get the `chip-build-mbedos` image in docker hub sooner than latter this will help with Github action files. 

At the same time I would be in favor of merging this PR now as the build environment is complete in `chip-build-mbedos` and complete the vscode integration in another PR. 

What do you think ?

**Comment by pan-** [comment]
Side note, @lukgni do you know why python packages are reinstalled with the bootstrap operation ? they are already present in `~/.local/bin`.

**Comment by lukgni** [comment]
Both scripts (script/bootstrap.sh and script/activate.sh) are part of [pigweed](https://pigweed.dev/docs/getting_started.html) tool. 

- **script/bootstrap.sh** is used to pull modules defined in script/requirements.txt (python) and script/pigweed.json (system). All of them will be installed in internal pigweed directory.
- **script/activate.sh** configures virtualenv to start using modules installed during bootstrap process. 

Pigweed doesn't care about tools like gcc, cmake already installed in the system because at the end it will configure environment to use tools from its own directory.
Eeach vscode task is executed in new terminal shell. As @ATmobica mentioned on our daily call: only nRF-specific vscode tasks use activate.sh script to setup environment.

**Comment by lukgni** [comment]
@pan- I submitted  [PR](https://github.com/project-chip/connectedhomeip/pull/4620)  on the upstream repo with changes only associated with creating chip-build-mbedos image.

**Comment by lukgni** [comment]
For now we should only care about adding chip-build-mbed-os image. We have to clarify couple of things first (especially with pigweed) before we could take next steps.
For that reason I will close this ticket and create a new one which will be similiar to https://github.com/project-chip/connectedhomeip/pull/4620).

**Pull request closed without merging**